Support read/write dynamic fields, check difference before updating

sky4star opened the pull request:
1. support dynamic fields, you can add any new fields into map named
   properties, it will be flattened as {“field1”:”value1”, …} when
   persisting into Elasticsearch.
2. performance tuning, not update a doc if the content doesn’t change
3. update pom

sky4star committed:
Support read/write dynamic fields, check difference before updating

1. support dynamic fields, you can add any new fields into map named
properties, it will be flattened as {“field1”:”value1”, …} when
persisting into Elasticsearch.
2. performance tuning, not update a doc if the content doesn’t change
3. update pom

wezhang commented on pom.xml:
```diff
@@ -29,37 +23,12 @@
-        <license>
-            <name>The Apache Software License, Version 2.0</name>
-            <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
-        </license>
```
这个不能删掉……

wezhang commented on pom.xml:
```diff
@@ -4,13 +4,7 @@
-    </parent>
-
-    <groupId>com.github.fhuss</groupId>
+    <groupId>com.lambdacloud.streaming</groupId>
```
com.lambdacloud.realtime.streaming

wezhang commented on pom.xml:
```diff
@@ -79,7 +48,6 @@
             <groupId>org.apache.storm</groupId>
             <artifactId>storm-core</artifactId>
             <version>${storm.version}</version>
-            <scope>provided</scope>
```
这个不能去掉

sky4star replied:
这个不去掉本地的编译运行会有问题，我可以先不去掉，你到时候帮我看下

wezhang commented on pom.xml:
```diff
@@ -4,13 +4,7 @@
-    <groupId>com.github.fhuss</groupId>
+    <groupId>com.lambdacloud.streaming</groupId>
     <artifactId>storm-elasticsearch</artifactId>
     <version>0.3.0</version>
```
版本号改成0.3.0.1

wezhang commented on src/main/java/com/github/fhuss/storm/elasticsearch/state/EnCachedMap.java:
```diff
@@ -0,0 +1,80 @@
+package com.github.fhuss.storm.elasticsearch.state;
```
加上咱们公司的copyright

wezhang commented on src/main/java/com/github/fhuss/storm/elasticsearch/state/EnCachedMap.java:
```diff
@@ -0,0 +1,80 @@
+ * unchanged state
+ * @author sky4star
+ */
+public class EnCachedMap<T> implements IBackingMap<T> {
```
En是enhanced的意思吗？那EnhancedCachedMap或者ExtCachedMap, CachedMapPlus会更清晰一些，根据这个类的行为可能CachedMapPutOnUpdate会更贴切些

wezhang replied:
另外我看对CachedMap只改动了multiPut，那么继承CachedMap然后只重载multiPut不好一些吗？

sky4star replied:
嗯有道理，好的

sky4star replied:
试了下不行，因为CachedMap有些必要的对象是私有的，即成之后没法访问

wezhang commented on src/main/java/com/github/fhuss/storm/elasticsearch/state/JsonDynamicFields.java:
```diff
@@ -0,0 +1,38 @@
+package com.github.fhuss.storm.elasticsearch.state;
```
我在想，也许我们不需要这个改动，只是在对象中加入一个Map()成员如何？

sky4star replied:
Map 对象反序列化出来的话，会变成{"mapname":{"f1":"v1", "f2":"v2", "f3":"v3"}} 但是咱们想要的是 {"fieldname1":"v1", "fieldname2": "v2"}

sky4star committed:
resolve comments

sky4star committed:
fix an issue, ExtCachedMap should return a clone

If ExtCachedMap returns the reference, and it is been updated, multiPut
function will always think the new document and the old one is the same
(since both use the same reference)
To make the logic simple, we only clone for our own JsonDynamicFields
